src/cluster.c

```diff
@@ -3659,7 +3659,8 @@ void clusterCron(void) {
         clusterNode *node = dictGetVal(de);
         /* Resize the send buffer if it is wasting
         * enough space. */
-        if (node->link != NULL && sdsavail(node->link->sndbuf) > SNDBUF_AVAILABLE_THRESHOLD) {
+        /* If available space is more than twice the size of the buffer then free up unused space */
+        if (node->link != NULL && sdsavail(node->link->sndbuf) / 2 > sdslen(node->link->sndbuf)) {
```

Comment:
because sdsMakeRoomFor can allocate twice of what's needed, i'm afraid that shrinking here when we have a 2x factor can create a resonance effect. that's why i suggested 4x.
@yoav-steinberg when you said the current code looks good, did you consider the concerns i posted in my first review?

Reply:
@oranagra I tend to agree that we should be on the save side. Because `sdsavail` takes into account internal fragmentation we might actually have a situation where `sdsMakeRoomFor` allocates even more than twice the size. Then then we might go back and forth shrinking and "making room for". So lets go with the extra safety margin of `/ 4` instead.
```

Reply:
usually `sdslen(node->link->sndbuf)` always return 0 , so `sdsRemoveFreeSpace` will be excuted in each cron and `link->sndbuffer` will be freed and malloc frequently。
What do you think that adding a hard limit before resize the buffer? it looks like 
`sdsalloc(node->link->sndbuf) > PROTO_IOBUF_LEN && sdsavail(node->link->sndbuf) / 4 > sdslen(node->link->sndbuf)`

Reply:
I think that after a single call to `sdsRemoveFreeSpace()` we won't have any available space anymore, so if `sndbuf` has a len of 0 and no more available space we won't call `sdsRemoveFreeSpace()` anymore. Also `PROTO_IOBUF_LEN` isn't **that** small, and in many cases we don't need such a buffer. So I don't see any harm in the current implementation, it'll shrink to 0 and stay that way until we need to grow it.

Reply:
@yoav-steinberg  thanks for your words, I also think that  `PROTO_IOBUF_LEN` is a little expensive for most scenarios. in my test, all `node->link->sendbuf` shrink to 0 and immediately grow up to 8kB for a ping messages, this occurs in every cron, so I think we can optimize it a little to avoid frequent applications and free memory. I look forward to your opinion.

Reply:
I think the idea is that cron is called infrequently enough so that it can assume that if we can shrink it's worth shrinking since there will be a long interval until the next usage of the buffer. If we feel that the cron is "fighting" with the whoever it writing to the buffer in such a way that:
1. Each time we write cron comes along and shrinks it so the next write needs an allocation.
2. And each time cron frees some memory we write immediately after that requiring to re-allocate what was just freed.
Then the problem is a timing issue more that anything else. In such a case we need to increase the cron interval. Maybe we can try to call `clusterNodeCronResizeBuffers()` with `run_with_period(500)` or so be default.

Reply:
I'm not too familiar with cluster, but i guess there are cases where cluster bus traffic could be scarce (maybe same rate as cron)?
in clientsCronResizeQueryBuffer we now have some logic to avoid freeing the available space, if it's not large (at least 4kb), i.e. we know we could save some memory, but the extra malloc calls are not worth the effort.
maybe we should do something similar here?

Reply:
@oranagra  I agree with you. may I ask the reason of  number '4kb'

Reply:
sorry for the late reply (message lost in other notifications).
i don't know the reason Salvatore chose 4k in clientsCronResizeQueryBuffer, probably just that this means the memory loss / gain is not very high.